src/lib/log.c

```diff
@@ -31,15 +31,57 @@ void log_init(void)
 }
 
 
+/* Remove tty escapes for colors from buf.  */
+static void remove_escapes()
+{
+	bool in_escape = false;
+	char *dst = buf;
+	char *src = buf;
+
+	while (true) {
+		if ((src[0] == '\x1b') && (src[1] == '[')) {
+			in_escape = true;
+			src += 2;
+		} else if ((src[0] == 'm') && in_escape)     {
+			in_escape = false;
+			src += 1;
+		} else if (in_escape)     {
+			src += 1;
+		} else     {
+			*dst = *src;
+			if (!*dst) {
+				break;
+			}
+			dst += 1;
+			src += 1;
+		}
+	}
+}
+
+
+/* Whether str starts with pfx.  */
+static bool has_prefix(const char *str, const char *pfx)
+{
+	return (strncmp(str, pfx, strlen(pfx)) == 0);
+}
+
+
 /* Send the content of the buffer to the logger. */
 static void log_flush(void)
 {
 	buf[buf_idx] = 0;
-	if (strcasestr(buf, "cannot")
-		|| strcasestr(buf, "could not")
-		|| strcasestr(buf, "error")
-		|| strcasestr(buf, "failed")
-		|| strcasestr(buf, "unable to")) {
+	remove_escapes();
+	if (has_prefix(buf, "INFO[")) {
+		log_info("%s", buf+4);
+	} else if (has_prefix(buf, "WARN[")) {
+		log_warn("%s", buf+4);
+	} else if (has_prefix(buf, "ERRO[")) {
+		log_error("%s", buf+4);
+	} else if (strcasestr(buf, "cannot") ||
+	    strcasestr(buf, "could not") ||
+	    strcasestr(buf, "error") ||
+	    strcasestr(buf, "failed") ||
+	    strcasestr(buf, "unable to")) {
```

Comment:
Isn't `buf` the guest console log? How can you know enough about its structure and contents (in general) to assert that hardcoding all these heuristics makes sense at the Hyperkit level?

Reply:
Hi Ian!
There's no guarantee, it's purely a heuristics.  Log level are purely "decorators", they don't trigger actions as Panicf would do in go.  Log levels help drawing the attention to some events, and also decide whether the event is volatile or persistent.
So I'd rather have maybe _some_ false positive, than flat logs with a single level.
Given that we don't want massive changes in hyperkit, that's a reasonable heuristics.

Reply:
I'm afraid we cannot accept these sorts of application specific heuristics into a low level and generic component such as Hyperkit, so this will have to go.